lib/Backend/Lower.cpp

```diff
@@ -22547,6 +22547,11 @@ bool Lowerer::GenerateFastBooleanAndObjectEqLikely(IR::Instr * instr, IR::Opnd *
             }
         }
     }
+    else if (src1->GetValueType().IsLikelySymbol() && src2->GetValueType().IsLikelySymbol())
```

Comment:
Is this valid? I didn't think you could just compare by pointer?

Reply:
What are we comparing here that src1 and src2 are both symbols right? what if they are different symbols?

---
In reply to: [176284445](https://github.com/Microsoft/ChakraCore/pull/4857#discussion_r176284445) [](ancestors = 176284445)

Reply:
I think this is saying: if they are both symbols then we can compare the pointer values for equality

Reply:
Can we have tests please.

---
In reply to: [176284781](https://github.com/Microsoft/ChakraCore/pull/4857#discussion_r176284781) [](ancestors = 176284781)

Reply:
Discussed with @boingoing offline, cant seem to find a case where pointer comparison is not sufficient yet (Even the cross context case, the global symbol cache returns the same symbol from a cache).

Reply:
I tried to reuse existing JavascriptSymbol instances whenever possible so we could keep hitting the same PolymorphicInlineCache, so I'm happy to see that it's useful in other places too. If we're now claiming that Symbol reference comparison is valid and correct, we should also update some other places (this should both help perf and give us extra validation that this change is correct):

- In JavascriptOperators::StrictEqual, JavascriptConversion::SameValueCommon, JavascriptSymbol::Equals, and RecyclableObject::Equals should all assert that GetValue() compares the same as references in the symbol-symbol case, and return the result of comparing references
- In SameValueComparer::GetHashCode, remove the special case for Symbols
- In Map and Set implementations, let's include Symbols in the simple case since they don't need custom comparison any more. Important bit is TryCanonicalizeAsSimpleVar; also update comments in SetKind and MapKind enums.

I don't understand the cross-context case very well. If, in a parent frame, we do something like `Symbol.replace === someIframe.contentWindow.Symbol.replace`, does the Symbol for the other frame get marshalled to the current ScriptContext? If so, then I agree the cache should ensure reference equality.

Worth mentioning, but I think irrelevant for correct comparisons: a single Symbol can be allocated in multiple locations in the same ScriptContext, just not at the same time. If a Symbol is not strong-referenced but its PropertyId is used as a key in an object, or if its PropertyRecord exists in the global Symbol.for() registry, then the Symbol instance could be collected, and we'd create a new one when we iterate property keys or fetch it from Symbol.for().

---
In reply to: [176292138](https://github.com/Microsoft/ChakraCore/pull/4857#discussion_r176292138) [](ancestors = 176292138)

Reply:
@sethbrenith Thanks. I am running our internal tests locally for some of the assertions you also mentioned. It'll be good to do comparison changes across if we can conclude pointer comparison is sufficient.

Reply:
> I don't understand the cross-context case very well. If, in a parent frame, we do something like Symbol.replace === someIframe.contentWindow.Symbol.replace, does the Symbol for the other frame get marshalled to the current ScriptContext? If so, then I agree the cache should ensure reference equality.

Do you mean :

Reply:
Symbol.replace is an internal symbol which might have slightly different behavior vs a user-defined symbol or symbol defined by Symbol.for. But I think this is testing what we want to test.

Reply:
I guess Symbol.species might be a bit better because Symbol.replace doesn't exist by default. Anyhow, I stepped through that code and verified that it calls JavascriptSymbol::CloneToScriptContext when fetching the property from the cross-site object, so it seems to be behaving as expected.

---
In reply to: [176569623](https://github.com/Microsoft/ChakraCore/pull/4857#discussion_r176569623) [](ancestors = 176569623)